File: teams-ai-kit/python/samples/dev_server.py
"""Shared dev server for the Python samples (aiohttp): bot endpoint + reply loopback."""

from __future__ import annotations

from typing import Any, List

from aiohttp import web
from aiohttp.web_request import Request
from aiohttp.web_response import Response
from botbuilder.core import BotFrameworkAdapter, BotFrameworkAdapterSettings, TurnContext
from botbuilder.schema import Activity

from teams_ai_kit.application import App


def run_dev_server(app: App, name: str, model: str, port: int) -> None:
    replies: List[Any] = []
    # Anonymous adapter (empty app id/password) — same local behavior as the Node kit.
    adapter = BotFrameworkAdapter(BotFrameworkAdapterSettings(app_id="", app_password=""))

    async def messages(request: Request) -> Response:
        activity = Activity().deserialize(await request.json())

        async def logic(ctx: TurnContext) -> None:
            await app.on_turn(ctx)

        await adapter.process_activity(activity, "", logic)  # empty auth header = anonymous
        return Response(status=200)

    async def loopback(request: Request) -> Response:
        replies.append(await request.json())
        return web.json_response({"id": str(len(replies))})

    async def get_replies(request: Request) -> Response:
        out, replies[:] = replies[:], []
        return web.json_response(out)

    async def health(request: Request) -> Response:
        return web.json_response({"ok": True, "model": model})

    server = web.Application()
    server.router.add_post("/api/messages", messages)
    server.router.add_post("/v3/conversations/{conversation_id}/activities", loopback)
    server.router.add_post("/v3/conversations/{conversation_id}/activities/{activity_id}", loopback)
    server.router.add_get("/replies", get_replies)
    server.router.add_get("/health", health)

    print(f"{name} listening on http://localhost:{port}/api/messages (model: {model})")
    web.run_app(server, host="0.0.0.0", port=port, print=None)

File: teams-ai-kit/node/samples/dev_server.ts
/**
 * Dev server for samples: exposes POST /api/messages (the bot endpoint) plus
 * a local loopback the Bot Framework connector uses to deliver the bot's
 * replies — so `samples/client.ts` (which sets serviceUrl to this origin)
 * can fetch them via GET /replies. No registration, emulator, or tunnel needed.
 */

import express from "express";
import type { App } from "../src";

export function startDevServer(app: App, port: number, name: string, model: string): void {
  const server = express();
  const replies: any[] = [];

  server.use(express.json());
  server.get("/health", (_req, res) => res.json({ ok: true, model }));
  server.post("/api/messages", (req, res) => app.processActivity(req, res));
  // Connector loopback: the adapter delivers replies here; we keep them for /replies.
  server.post(["/v3/conversations/:conversationId/activities", "/v3/conversations/:conversationId/activities/:activityId"], (req, res) => {
    replies.push(req.body);
    res.status(200).json({ id: String(replies.length) });
  });
  server.get("/replies", (_req, res) => res.json(replies.splice(0)));

  server.listen(port, () => {
    console.log(`${name} listening on http://localhost:${port}/api/messages (model: ${model})`);
    console.log(`local test: npx tsx samples/client.ts --url http://localhost:${port}/api/messages "hello"`);
  });
}
